File: 2addTwoNumbers/addTwoNumbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        final var list = new LinkedList<Integer>();
        var carry = 0;
        while (Objects.nonNull(l1) || Objects.nonNull(l2) || carry != 0)
        {
            if (Objects.nonNull(l1))
            {
                carry += l1.val;
                l1 = l1.next;
            }
            if (Objects.nonNull(l2))
            {
                carry += l2.val;
                l2 = l2.next;
            }
            list.offer(carry % 10);
            carry /= 10;
        }
        final var result = new ListNode(0);
        var current = result;
        for (final var $: list)
        {
            current.next = new ListNode($);
            current = current.next;
        }
        return result.next;
    }
}

File: 2addTwoNumbers/addTwoNumbers.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function(l1, l2) {
    const result = new ListNode(null)
    let current = result
    let carry = 0
    while (l1 || l2 || carry)
    {
        if (l1)
        {
            carry += l1.val
            l1 = l1.next
        }
        if (l2)
        {
            carry += l2.val
            l2 = l2.next
        }
        current.next = new ListNode(carry % 10)
        carry = Math.trunc(carry / 10)
        current = current.next
    }
    return result.next
}
